ICE building futures-rs-all-test from rustc-benchmarks

A recent change has caused rustc to ICE on futures-rs-all-test from the rustc-benchmarks suite. I had a rustc checkout from September 26 that worked, and then I updated to tip of master on October 3 and the problem started.

The output, and the build command, are below. It looks like the `unwrap` call in `local_def_id` is panicking.

```
$ RUST_BACKTRACE=1 RUSTC=/home/njn/moz/rust0/x86_64-unknown-linux-gnu/stage1/bin/rustc make
cargo rustc --test all -- 
   Compiling log v0.3.6
   Compiling futures v0.1.0 (file:///home/njn/moz/rustc-benchmarks/futures-rs-test-all)
error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: run with `RUST_BACKTRACE=1` for a backtrace

thread 'rustc' panicked at 'called `Option::unwrap()` on a `None` value', src/libcore/option.rs:323
stack backtrace:
   1:     0x7f6d27216538 - std::sys::backtrace::tracing::imp::write::he09d58d1497100a9
                        at /home/njn/moz/rust0/src/libstd/sys/unix/backtrace/tracing/gcc_s.rs:42
   2:     0x7f6d271f4a8f - std::panicking::default_hook::{{closure}}::hf3f7aa526824ae88
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:247
   3:     0x7f6d271f398a - std::panicking::default_hook::h691f7ee16f920a74
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:257
   4:     0x7f6d271f3fd7 - std::panicking::rust_panic_with_hook::haa794968064098e8
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:451
   5:     0x7f6d271f3e64 - std::panicking::begin_panic::h78b83f343fc429be
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:413
   6:     0x7f6d271f3d89 - std::panicking::begin_panic_fmt::h8b3207332f403a68
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:397
   7:     0x7f6d271f3d17 - rust_begin_unwind
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:373
   8:     0x7f6d2723c63d - core::panicking::panic_fmt::h5cca666ac160c125
                        at /home/njn/moz/rust0/src/libcore/panicking.rs:69
   9:     0x7f6d2723c574 - core::panicking::panic::h30a00c8dd89101b4
                        at /home/njn/moz/rust0/src/libcore/panicking.rs:49
  10:     0x7f6d242e6294 - rustc::hir::map::definitions::Definitions::local_def_id::hec5ea47d4293d9af
                        at /home/njn/moz/rust0/src/libcore/macros.rs:21
                        at /home/njn/moz/rust0/src/librustc/hir/map/definitions.rs:258
  11:     0x7f6d26795014 - <rustc_resolve::build_reduced_graph::BuildReducedGraphVisitor<'a, 'b> as syntax::visit::Visitor>::visit_trait_item::hacc123954c682e8a
                        at /home/njn/moz/rust0/src/librustc_resolve/build_reduced_graph.rs:609
  12:     0x7f6d26794851 - syntax::visit::walk_item::h1f5fc20bc60b2f81
                        at /home/njn/moz/rust0/src/libsyntax/visit.rs:140
  13:     0x7f6d26794700 - syntax::visit::walk_item::h1f5fc20bc60b2f81
                        at /home/njn/moz/rust0/src/librustc_resolve/build_reduced_graph.rs:588
                        at /home/njn/moz/rust0/src/libsyntax/visit.rs:140
                        at /home/njn/moz/rust0/src/libsyntax/visit.rs:59
                        at /home/njn/moz/rust0/src/libsyntax/visit.rs:256
  14:     0x7f6d26794df0 - <rustc_resolve::build_reduced_graph::BuildReducedGraphVisitor<'a, 'b> as syntax::visit::Visitor>::visit_item::hcd12b1ea6bf07c84
                        at /home/njn/moz/rust0/src/librustc_resolve/build_reduced_graph.rs:588
  15:     0x7f6d26760c3e - rustc_resolve::macros::<impl syntax::ext::base::Resolver for rustc_resolve::Resolver<'a>>::visit_expansion::h14f796d72ffc1ff0
                        at /home/njn/moz/rust0/src/libsyntax/ext/expand.rs:92
                        at /home/njn/moz/rust0/src/librustc_resolve/macros.rs:76
  16:     0x7f6d20a83523 - syntax::ext::expand::MacroExpander::collect_invocations::hfc747101b6389c1c
                        at /home/njn/moz/rust0/src/libsyntax/ext/expand.rs:282
  17:     0x7f6d20a809f2 - syntax::ext::expand::MacroExpander::expand::h5f295468aa4384c0
                        at /home/njn/moz/rust0/src/libsyntax/ext/expand.rs:221
  18:     0x7f6d20a80311 - syntax::ext::expand::MacroExpander::expand_crate::h28e4d136072e55c4
                        at /home/njn/moz/rust0/src/libsyntax/ext/expand.rs:201
  19:     0x7f6d20a93460 - syntax::ext::expand::expand_crate::hd3de49be73535dde
                        at /home/njn/moz/rust0/src/libsyntax/ext/expand.rs:873
  20:     0x7f6d2750c10c - rustc_driver::driver::phase_2_configure_and_expand::{{closure}}::hec6c2fd69fe26113
                        at /home/njn/moz/rust0/src/librustc_driver/driver.rs:689
  21:     0x7f6d274fb34d - rustc_driver::driver::phase_2_configure_and_expand::h6ae2e19d626066fe
                        at /home/njn/moz/rust0/src/librustc/util/common.rs:38
                        at /home/njn/moz/rust0/src/librustc_driver/driver.rs:654
  22:     0x7f6d274f317d - rustc_driver::driver::compile_input::h0271fc4a3333b327
                        at /home/njn/moz/rust0/src/librustc_driver/driver.rs:121
  23:     0x7f6d27597e0d - rustc_driver::run_compiler::h1315ba568dec0ab8
                        at /home/njn/moz/rust0/src/librustc_driver/lib.rs:224
  24:     0x7f6d2758bbd8 - std::panicking::try::do_call::h10f47cce5c9324d1
                        at /home/njn/moz/rust0/src/librustc_driver/lib.rs:1145
                        at /home/njn/moz/rust0/src/librustc_driver/lib.rs:137
                        at /home/njn/moz/rust0/src/librustc_driver/lib.rs:1079
                        at /home/njn/moz/rust0/src/libstd/panic.rs:255
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:356
  25:     0x7f6d272279a7 - __rust_maybe_catch_panic
                        at /home/njn/moz/rust0/src/libpanic_unwind/lib.rs:97
  26:     0x7f6d27593c98 - <F as alloc::boxed::FnBox<A>>::call_box::ha787c4c8ae12fdd8
                        at /home/njn/moz/rust0/src/libstd/panicking.rs:332
                        at /home/njn/moz/rust0/src/libstd/panic.rs:311
                        at /home/njn/moz/rust0/src/libstd/thread/mod.rs:277
                        at /home/njn/moz/rust0/src/liballoc/boxed.rs:595
  27:     0x7f6d271fa483 - std::sys_common::thread::start_thread::h5c65a2bf0ace983e
                        at /home/njn/moz/rust0/src/liballoc/boxed.rs:605
                        at /home/njn/moz/rust0/src/libstd/sys/common/thread.rs:21
  28:     0x7f6d271ff7e8 - std::sys::thread::Thread::new::thread_start::h2d239c26e84b740e
                        at /home/njn/moz/rust0/src/libstd/sys/unix/thread.rs:85
  29:     0x7f6d1ff126f9 - start_thread
  30:     0x7f6d26ed2b5c - clone
  31:                0x0 - <unknown>

error: Could not compile `futures`.

To learn more, run the command again with --verbose.
makefile:4: recipe for target 'all' failed
make: *** [all] Error 101
```

--- a/src/librustc_resolve/build_reduced_graph.rs
+++ b/src/librustc_resolve/build_reduced_graph.rs
@@ -605,6 +605,10 @@ impl<'a, 'b> Visitor for BuildReducedGraphVisitor<'a, 'b> {
         let parent = self.resolver.current_module;
         let def_id = parent.def_id().unwrap();
 
+        if let TraitItemKind::Macro(_) = item.node {
+            return self.visit_invoc(item.id);
+        }
+
         // Add the item to the trait info.
         let item_def_id = self.resolver.definitions.local_def_id(item.id);
         let mut is_static_method = false;
@@ -615,7 +619,7 @@ impl<'a, 'b> Visitor for BuildReducedGraphVisitor<'a, 'b> {
                 (Def::Method(item_def_id), ValueNS)
             }
             TraitItemKind::Type(..) => (Def::AssociatedTy(item_def_id), TypeNS),
-            TraitItemKind::Macro(_) => return self.visit_invoc(item.id),
+            TraitItemKind::Macro(_) => bug!(),  // handled above
         };
 
         self.resolver.trait_item_map.insert((item.ident.name, def_id), is_static_method);

